psoc/grobot_workspace/grobot_module_controller.cydsn/dht11.c

```diff
@@ -0,0 +1,234 @@
+    decoded_humidity_data=(uint8_t)((humidity_values[7]<<7) + (humidity_values[6]<<6) + (humidity_values[5]<<5) + (humidity_values[4]<<4) + (humidity_values[3] <<3) + (humidity_values[2] <<2) + (humidity_values[1]<<1) + humidity_values[0]);
+    
+    //Now the data is decoded
+    decode_flag = 1;
+    //From here we will have the acctual values. We can call functions from
+    //here or send it out depending on what we decide to go with 
+}    
+
+//Name: Delay_Timer_ISR_Handler
+//Summary:
+//    Interrupt Service Routine. Check the Delay_Timer status and clear the interrupt
+//    change the Count and Executed status
+
+_CY_ISR(Delay_Timer_ISR_Handler){
+    
+    if(Delay_Timer_ReadPeriod() == DELAY_ACQUISITION){
+        count=2;                  
+        executed=0;             
+    }else if(Delay_Timer_ReadPeriod() == DELAY_PROCESSING){ 
+        decode_flag=1;
+    }
+    
+    //Write 1 to the Contro_Reg_Delay_Timer to disable the Timer function 
+    Control_Reg_Delay_Timer_Write(1);
+    //Read the status register of Delay_Timer to clear interrupt
+    Delay_Timer_ReadStatusRegister();
+    //Clear the Delay_Timer_ISR interrupt
+    Delay_Timer_ISR_ClearPending();
+}    
+
+_CY_ISR(DHT_Pin_ISR_Handler){
+    //Clear the interrupt request of the ISR DHT_Pin_ISR 
+    DHT_Pin_ISR_ClearPending();
+    //Clear the interrupt request of the pin DHT_Pin
+    DHT_Pin_ClearInterrupt();
+    //Read the counter values of Data_Timer for processing
+    countvalues_datatimer[risingedge_counts]=Data_Timer_ReadCounter();
+    risingedge_counts++;
+}
+
+void dht_init(){
+    check = 0;
+    decode_flag = 0;
+    
+    //Starts the data timer and delay timer components
+    Data_Timer_Start();
+    Delay_Timer_Start();
+    
+    //Write Controlregdatatimer and controlregdealytimer value to 1 to keep them reset
+    Control_Reg_Data_Timer_Write(1);
+    Control_Reg_Delay_Timer_Write(1);
+    
+    //The pin value will be written as 1 as the starting condition
+    DHT_Pin_Write(1);
+    
+    //Initilize dhtpinisr and delaytimeriser
+    DHT_Pin_ISR_StartEx(DHT_Pin_ISR_Handler);
+    Delay_Timer_ISR_StartEx(Delay_Timer_ISR_Handler);
+    
+    //Disable the ISR compondents
+    DHT_Pin_ISR_Disable();
+    Delay_Timer_ISR_Disable();
+    
+    //Initilize count to 0 and executed to 1
+    count = 0;
+    executed = 1;
+}   
+
+
+//What dht_start does:
+//    Starts the dht components
+//    Get output from dht11
+//    Send out results
+//
+//Paramers: None
+//Returns: None
+
+
+void dht_start(){
+    
+    // Enable Global Interrupts
+    CyGlobalIntEnable;
+    
+    //This is going to run until decode dht data is called
+    //After this is done, the loop will end, and the data will
+    //be in the respective variables
+    for(;;){
```

Comment:
Okay, this is closer to the right idea. However, I don't want to sit in a loop at all. I want to call ```dht_start```, and then do other things for awhile. Then, at some arbitrary point in the future, I want to be able to call ```get_decoded_<whatever>```, and have it give me the value that it read immediately.

If you have a complex, multi-step process for reading the sensor, like you do here, the way you handle that is you have ```dht_start``` kick off the process and return immediately. Then, when the relevant interrupt fires, indicating that step 1 is done, you do whatever you need to do with the data, set up step 2, and return from the ISR. After the relevant interrupt for step 2 fires, you handle the data and set up step 3, and so on, until you're done reading the information, at which point you stick it in your global variables, and then you're completely finished. I realize that this is not entirely intuitive, so if you need me to explain it better on Slack or in person, just tell me.